### PR TITLE
feat: add 5-hour blocks report command

### LINKED ISSUE #23: 5 hours sessions tracker
### Discussed in https://github.com/ryoppippi/ccusage/discussions/20

<div type='discussions-op-text'>

<sup>Originally posted by **IntegralMedia01** June  7, 2025</sup>
please add ability to keep track of those 5 hours sessions. how many sessions already used for the month.</div>

### PR DESCRIPTION
**Pull request by a-c-m**
## Summary

This PR implements a new `blocks` command to analyze Claude Code usage in 5-hour billing windows as requested in #23.

## Features

- **5-Hour Block Tracking**: Groups usage into Claude's 5-hour session windows that start with first message
- **Active Block Monitoring**: Shows time elapsed/remaining and burn rate for current sessions
- **Token Limit Support**: Set limits with `-t <number>` or use `-t max` to automatically use highest previous block
- **Usage Projections**: Projects total usage if current burn rate continues for the full 5 hours
- **Recent Filter**: `--recent` flag shows only last 3 days of blocks
- **Detailed Active View**: `--active` flag shows detailed view of current active block

## Implementation Details

- Core logic in `src/utils/five-hour-blocks.ts` handles block detection and calculations
- New CLI command in `src/commands/blocks.ts` with table and JSON output formats
- Integrated with existing data loader infrastructure
- Added comprehensive documentation to README

## Example Output

```
┌────────────────────────────────────────────────────────┬─────────────┬─────────────────────────┬────────┬────────┬─────────┐
│ Block Start                                            │ Duration/   │ Models                  │ Tokens │ %      │ Cost    │
│                                                        │ Status      │                         │        │        │         │
├────────────────────────────────────────────────────────┼─────────────┼─────────────────────────┼────────┼────────┼─────────┤
│ 6/16/2025, 6:21:16 PM (1h 25m elapsed, 3h 35m remain) │ ACTIVE      │ claude-opus-4-20250514  │ 25,979 │ 41.8%  │ $74.54  │
├────────────────────────────────────────────────────────┼─────────────┼─────────────────────────┼────────┼────────┼─────────┤
│                      (assuming 62,081 token limit)    │ REMAINING   │                         │ 36,102 │ 58.2%  │         │
├────────────────────────────────────────────────────────┼─────────────┼─────────────────────────┼────────┼────────┼─────────┤
│                       (assuming current burn rate)    │ PROJECTED   │                         │ 91,443 │ 147.3% │ $262.09 │
└────────────────────────────────────────────────────────┴─────────────┴─────────────────────────┴────────┴────────┴─────────┘
```

## Testing

- All existing tests pass
- TypeScript compilation successful
- Manual testing with various scenarios including active blocks, token limits, and edge cases

Fixes #23

🤖 Generated with [Claude Code](https://claude.ai/code)

Co-Authored-By: @a-c-m

## Summary by CodeRabbit

- **New Features**
	- Introduced a new "blocks" command to the CLI, providing usage reports grouped by 5-hour billing windows.
	- Added detailed reporting for active and recent 5-hour blocks, including projections, token limits, and cost calculations.
	- Enhanced output options with JSON formatting and customizable display modes.
- **Documentation**
	- Updated the README with a comprehensive section on the new 5-hour blocks report, including usage examples and feature descriptions.

### CODE REVIEW

File: src/commands/blocks.ts [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+	if (hours > 0) {
+		return `${start} (${hours}h ${mins}m)`;
+	}
+	return `${start} (${mins}m)`;
+}
+
+function formatModels(models: string[]): string {
+	if (models.length === 0) { return '-'; }
+	if (models.length === 1) { return models[0] ?? '-'; }
+	return `${models[0]} +${models.length - 1}`;
```
Comment (dashed) [inline src/commands/blocks.ts]:
```suggestion
	return models.join('\n');
```

File: src/utils/five-hour-blocks.ts [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	let costUSD = 0;
+	const modelsSet = new Set<string>();
+
+	for (const entry of entries) {
+		tokenCounts.inputTokens += entry.usage.inputTokens;
+		tokenCounts.outputTokens += entry.usage.outputTokens;
+		tokenCounts.cacheCreationInputTokens += entry.usage.cacheCreationInputTokens;
+		tokenCounts.cacheReadInputTokens += entry.usage.cacheReadInputTokens;
+		costUSD += entry.costUSD || 0;
+		modelsSet.add(entry.model);
```
Comment (dashed) [inline src/utils/five-hour-blocks.ts]:
```suggestion
		// Skip synthetic model
		if (entry.model !== '<synthetic>') {
			modelsSet.add(entry.model);
		}
```

**Comment by dashed** [comment]
@a-c-m this works great! thanks! 

I left some minor feedback about being able to see all the models used within the 5 hour sessions blocks.

File: src/utils/five-hour-blocks.ts (hunk 2 above)
Comment (a-c-m) [reply src/utils/five-hour-blocks.ts]: Synthetic is actually useful to see, as it tends to be used when you hit your limit (i think)

Comment (dashed) [reply src/utils/five-hour-blocks.ts]: @a-c-m I don't really have strong opinions about it.

I suggested it for consistency with how it's filtered out:
- https://github.com/ryoppippi/ccusage/blob/486b74d155f404524d31ba2e30542ae0e33f430d/src/data-loader.ts#L141-L144
- https://github.com/ryoppippi/ccusage/blob/486b74d155f404524d31ba2e30542ae0e33f430d/src/data-loader.ts#L179-L182
- https://github.com/ryoppippi/ccusage/blob/486b74d155f404524d31ba2e30542ae0e33f430d/src/data-loader.ts#L295-L303
- https://github.com/ryoppippi/ccusage/blob/486b74d155f404524d31ba2e30542ae0e33f430d/src/data-loader.ts#L779-L782

Comment (ryoppippi) [reply src/utils/five-hour-blocks.ts]: @a-c-m could you tell me why you don't skip ` '<synthetic>'`

Comment (ryoppippi) [reply src/utils/five-hour-blocks.ts:132]: @ben-vargas can you review it?

**Comment by ryoppippi** [comment]
@ben-vargas could you review it? it looks brilliant

**Review by ryoppippi — APPROVED** [review]
LGTM
